### ISSUE TITLE
Is it missing some activation functions between some layers?

### ISSUE THREAD

**Issue by BridgetteSong** [facebookresearch/AudioDec#9]
Thanks for your work. I have trained model in my own dataset. I met same question as [ISSUE7](https://github.com/facebookresearch/AudioDec/issues/7). When I checked the model, I found some difference in AutoEncoder:

- Before encoder_output is feeded into Projector, is an activation function needed?
- Before ConvTranspose1d, is an activation function needed?
- add tanh activation function in Decoder final out?

In other popular implementations, they all added those. So I add those:

- add an activation function before https://github.com/facebookresearch/AudioDec/blob/9b498385890b38de048f2db535c2fbf8cbeea80b/models/autoencoder/modules/projector.py#L50
- add an activation function before https://github.com/facebookresearch/AudioDec/blob/9b498385890b38de048f2db535c2fbf8cbeea80b/models/autoencoder/modules/decoder.py#L62
- add an activation function before https://github.com/facebookresearch/AudioDec/blob/9b498385890b38de048f2db535c2fbf8cbeea80b/models/autoencoder/modules/decoder.py#L120
- add a tanh() after https://github.com/facebookresearch/AudioDec/blob/9b498385890b38de048f2db535c2fbf8cbeea80b/models/autoencoder/modules/decoder.py#L120

When I added those and trained again, I got some improvement in unseen datasets than your base when I only trained AutoEncoder with discriminators and don't finetune it with AudioDec.  
BTW, I trained model only with Librispeech and AIShell with 16K sampling_rate and tested model by another clean TTS dataset with training 160K steps. When my model is finished(total 800k), I will compare final results, upload some demos and share my training config.

**Comment by BridgetteSong** [comment]
demo with 200k steps
[demo.zip](https://github.com/facebookresearch/AudioDec/files/12157003/demo.zip)

**Comment by bigpon** [comment]
Hi,
Thanks for the interesting experiments! 
I think it is reasonable to add more nonlinearity to the model to enhance its modeling ability once the training is still stable. 
If you have more detailed results in any form (demo page, paper, etc.), please feel free to share them with us and I will update the README to show that adding activation functions will improve the robustness to unseen data.

**Comment by BridgetteSong** [comment]
@bigpon I confirmed changing autoencoder model can improve results. I changed as follows:

1. it is very important to add **activation functions** as I said above(**highly recommended**). I use **Snake activation function** not **ELU** or **LeakyReLU**
2. it is very important to add **weighnorm**, which can ensure training stability and model results significantly(**highly recommended**).
3. Appropriately increasing **code_dim** and **model size** can improve audio reconstruction quality(**recommended code_dim=128** but I use **256**)
4. I use **noncausal** training mode, **MPD** + **ComplexMRD** as discriminators, **MultiMelLoss**, trained by **AdamW** and **ExponentialLR**
5. BTW, there are some errors as your **MRD** because **intermediate convolution outputs** are **missed** and can't be computed **feature loss**.
https://github.com/facebookresearch/AudioDec/blob/9b498385890b38de048f2db535c2fbf8cbeea80b/models/vocoder/modules/discriminator.py#L568

Here is my training **config.yaml**
[symAD_librispeech_16000_hop160_base.txt](https://github.com/facebookresearch/AudioDec/files/12212631/symAD_librispeech_16000_hop160_base.txt)


**demos** for new config with training **200K steps** by using **librispeech** and **aishell** datasets

[demo.zip](https://github.com/facebookresearch/AudioDec/files/12212595/demo.zip)

**Comment by bigpon** [comment]
Hi @BridgetteSong,
- Thanks for the great efforts of investigation! I will check the results of 48kHz VCTK corpus.
- Do you have any plan to write a paper about your finding? If you write any paper, I will add the info to the README for others' reference.
- You are correct. The MRD actually has these problems, and I will fix them.
- Where do you put the WeightNorm layers?
- Could you also provide the results of the original AudioDec for references? (I assume the audiodec results in demo.zip are the modified version, right?)
- According to your conclusion, will these modifications increase the quality for arbitrary dataset? or the robustness of unseen dataset? Since you train and test the model using libritts and aishell, I assume that these modifications will increase the reconstruction quality for seen data, right?

**Comment by BridgetteSong** [comment]
@bigpon 

- I don't have the idea of ​​writing a paper yet, but I'm interested in improving the effect of Encodec.
- I add **WeightNorm** on **each Conv1d** layer like this: 
https://github.com/facebookresearch/AudioDec/blob/9b498385890b38de048f2db535c2fbf8cbeea80b/layers/conv_layer.py#L46
`self.conv = torch.nn.utils.weight_norm(nn.Conv1d(**))`
    
    I see that WeightNorm is added in 2nd stage of AudioDec by default, so **the difference is training 1st stage**. I think **stability of AutoEncoder** is very important to train 2nd stage or just only train one stage same as mine.
- I trained model by **libritts** and **aishell** dataset, but **test on an unseen dataset**(**audios** in **demo.zip** are **unseen** which are from **another TTS dataset**, even includes **a singing demo**), so these modifications can increase the quality for arbitrary dataset
- I can't provide  the results of the original AudioDec because the model is changed, the **demo.zip** is the modified version. but **demo.zip hifi dir** contains of the **16k and 24k original audios**, if someone has trained an original AudioDec model, just use **real audios in demo.zip** to test.

**Comment by bigpon** [comment]
Hi,
Thanks for your investigation!

According to our internal experiments, we get some conclusions.
1. Adding more activation functions like HiFiGAN will slightly increase the unseen data robustness. However, it is very similar to our 2-stage approach, which already used HiFiGAN as the decoder.
2. The snake activation doesn’t show marked improvements over the ELU activation. In some cases, the snake activation even achieves much worse speech quality. We think the instability of the snake activation might cause the problem.
3. Instand of adding activations, we found that increasing the bitrate to a reasonable scale (ex: 24kbps as Opus) will significantly improve the unseen data robustness, which somehow makes sense since it reduces the modeling difficulties. However, the very low bitrate feature is essential for some temporal-sensitive tasks such as LLM-based speech generation. Therefore, without greatly changing the architecture, adopting more training data will be a compromise. (We are investigating a new architecture for unseen data robustness and hope to release it soon.)

On the other hand, the 2D conv padding issue of the MSTFT discriminator has been fixed, and the corresponding models have been updated. Thanks for your contributions again.

**Comment by a897456** [comment]
Hi @BridgetteSong 
My experiment was similar to yours, using LibriSpeech to train AudioDec. In the first step, I used your config (very professional) to train autoencder, but there was always current sound. I don't understand why there is such a big gap with the demo you provided

**Comment by a897456** [comment]
> 2. The snake activation doesn’t show marked improvements over the ELU activation. In some cases, the snake activation even achieves much worse speech quality. We think the instability of the snake activation might cause the problem.

@bigpon Hi bigpon, 
I found that someone published a paper based on the soundstream framework, which mentioned the benefits of snake activation function as an innovation point, but this is not consistent with your conclusion, so I don't know what went wrong

![image](https://github.com/facebookresearch/AudioDec/assets/28752526/63134122-9f8f-49d6-a23b-45bb0fb7e35f)

**Comment by bigpon** [comment]
According to [DAC](https://github.com/descriptinc/descript-audio-codecl), they claimed that snake is much better.
However, in AudioDec architecture, we didn't find the tendency.
Two possible reasons,
1. Snake is sensitive to initialization and training, so we might not optimize the training processing of AudioDec with sanke (e.g. we didn't apply layer normalization, etc.)
2. Snake is better than Leaky ReLU but we use ELU here.

Since we gave it a quick try without carefully tuning the hyperparameters, further investigations are required.
